Print module

TurboGit opened the pull request:
This is a first version for review. I propose it a bit earlier than before as the gtk3 is now merged it is harder for me to keep this branch up-to-date with master.

As proposed by houz I'll be happy to get help to migrate this piece of code to Gtk3.

I'll be also happy to get comment from review to straighten the code where needed.

Thanks.

hanatos commented on src/common/cups_print.h:
```diff
@@ -0,0 +1,98 @@
+
+#define MAX_NAME 128
+
+typedef enum _alignement
```
not that it matters.. but this looks like a typo (french spelling?)

TurboGit replied:
indeed, a french spelling! fixed.

hanatos commented on src/common/printprof.c:
```diff
@@ -0,0 +1,763 @@
+
+#define SW "-"
+
+typedef struct _print_profile_data_t
```
i think you can actually give the struct the same name and the same tag. why the leading underscore?

TurboGit replied:
right, fixed here and in other places.

hanatos commented on src/control/jobs/control_jobs.c:
```diff
@@ -1033,6 +1034,50 @@ static int32_t dt_control_export_job_run(dt_job_t *job)
+    fprintf(stderr, "[print] printer profile %s and intent %d\n", pp->pinfo.printer.profile, pp->pinfo.printer.intent);
+    if (!dt_apply_printer_profile(filename, pp->pinfo.printer.profile, pp->pinfo.printer.intent))
+    {
+      dt_control_log(_("failed to apply printer profile '%'"), pp->pinfo.printer.profile);
```
missing format character after % ?

TurboGit replied:
good catch, %s of course here. fixed.

LebedevRI commented on CMakeLists.txt:
```diff
@@ -30,6 +30,7 @@ option(USE_OPENJPEG "Enable JPEG 2000 support" ON)
 option(USE_WEBP "Enable WebP export support" ON)
 option(BUILD_CMSTEST "Build a test program to check your system's color management setup" ON)
 option(USE_OPENEXR "Enable OpenEXR support" ON)
+option(USE_PRINT "Build the print module" ON)
```
I think it should be BUILD_PRINT, just like BUILD_SLIDESHOW

TurboGit replied:
Sure, fixed.

LebedevRI commented on src/libs/print_settings.c:
```diff
@@ -0,0 +1,1168 @@
+  // create papers combo as filled when adding printers
+  d->papers = GTK_COMBO_BOX(gtk_combo_box_text_new());
+
+  GList *printers = dt_get_printers();
```
I do not see where this is being freed (what was allocated in dt_get_printer_info())

TurboGit replied:
Now fixed also, I had forgotten it in the first batch of fixes.

LebedevRI commented on src/common/cups_print.h:
```diff
@@ -0,0 +1,98 @@
+  bottom_left, bottom, bottom_right
+} dt_alignment_t;
+
+typedef struct dt__paper_info_t
```
You do not need __ there, dt_paper_info_t should work just fine.

TurboGit replied:
Right, cut&paste error. Fixed.

LebedevRI commented on src/common/printprof.c:
```diff
@@ -0,0 +1,761 @@
+  TIFF *in, *out;
+} dt_print_profile_data_t;
+
+#define FatalError printf
```
Why not use printf() directly?
Alternatively, maybe you should call dt_control_log()

TurboGit replied:
Right, I have used fprintf on stderr with "error:" as prefix as in many places already. will help dev to track down errors.

LebedevRI commented on src/control/jobs/control_jobs.c:
```diff
@@ -1033,6 +1034,50 @@ static int32_t dt_control_export_job_run(dt_job_t *job)
+  GList *t = params->index;
+  dt_control_print_t *pp = (dt_control_print_t*)params->data;
+  int32_t imgid = 0;
+  char filename[PATH_MAX];
```
= { 0 }

TurboGit replied:
Fixed.

LebedevRI commented on src/control/jobs/control_jobs.c:
```diff
@@ -1033,6 +1034,50 @@ static int32_t dt_control_export_job_run(dt_job_t *job)
+
+  if (t) imgid = GPOINTER_TO_INT(t->data);
+
+  g_strlcpy(filename, pp->filename, PATH_MAX);
```
sizeof(pp->filename)

TurboGit replied:
Fixed.

LebedevRI commented on src/control/jobs/control_jobs.c:
```diff
@@ -1033,6 +1034,50 @@ static int32_t dt_control_export_job_run(dt_job_t *job)
+  if (t) imgid = GPOINTER_TO_INT(t->data);
+
+  g_strlcpy(filename, pp->filename, PATH_MAX);
+  g_strlcat(filename, ".tif", PATH_MAX);
```
sizeof(filename)

TurboGit replied:
Fixed too.

LebedevRI commented on src/control/jobs/control_jobs.c:
```diff
@@ -1033,6 +1034,50 @@ static int32_t dt_control_export_job_run(dt_job_t *job)
+
+  char tag[256] = { 0 };
+  guint tagid = 0;
+  snprintf (tag, 256, "darktable|printed|%s", pp->pinfo.printer.name);
```
sizeof(tag)

TurboGit replied:
Fixed.

LebedevRI commented on src/control/jobs/control_jobs.c:
```diff
@@ -1319,6 +1364,56 @@ void dt_control_export(GList *imgid_list, int max_width, int max_height, int for
+  data->max_height = max_height;
+  data->format_index = format_index;
+  data->storage_index = storage_index;
+  g_strlcpy(data->filename, filename, PATH_MAX);
```
sizeof()

TurboGit replied:
Fixed.

LebedevRI commented on src/control/jobs/control_jobs.c:
```diff
@@ -1319,6 +1364,56 @@ void dt_control_export(GList *imgid_list, int max_width, int max_height, int for
+  }
+  dt_control_job_set_params(job, params);
+  params->index = imgid_list;
+  dt_control_print_t *data = (dt_control_print_t*)malloc(sizeof(dt_control_print_t));
```
if(!data)

TurboGit replied:
Good catch! Fixed.

LebedevRI commented on src/libs/print_settings.c:
```diff
@@ -0,0 +1,1168 @@
+
+  // the exported filename, on /tmp we do not want to mess with user's home directory
+
+  char filename[PATH_MAX];
```
= {0};

TurboGit replied:
Fixed.

LebedevRI commented on src/libs/print_settings.c:
```diff
@@ -0,0 +1,1168 @@
+
+  char filename[PATH_MAX];
+
+  snprintf(filename, PATH_MAX, "/tmp/pf_%d", imgid);
```
sizeof()

TurboGit replied:
Fixed.

LebedevRI commented on src/libs/print_settings.c:
```diff
@@ -0,0 +1,1168 @@
+{
+  dt_lib_print_settings_t *d = (dt_lib_print_settings_t*)malloc(sizeof(dt_lib_print_settings_t));
+  self->data = d;
+  self->widget = gtk_table_new(18, 2, FALSE);
```
GTK3: grid

TurboGit replied:
I'll let the expert convert to GTK3. houz has proposed to do that.

LebedevRI commented on src/libs/print_settings.c:
```diff
@@ -0,0 +1,1168 @@
+  //// alignments
+
+  // Create the 3x3 gtk table toggle button table...
+  GtkTable *bat = GTK_TABLE(gtk_table_new(3,3,TRUE));
```
GTK3: grid

LebedevRI commented on src/libs/print_settings.c:
```diff
@@ -0,0 +1,1168 @@
+  d->profiles = g_list_append(d->profiles, prof);
+
+  // read datadir/color/out/*.icc
+  char datadir[PATH_MAX];
```
= { 0 };

TurboGit replied:
Fixed.

LebedevRI commented on src/views/view.c:
```diff
@@ -69,7 +69,7 @@ void dt_view_manager_init(dt_view_manager_t *vm)
                       "map",
 #endif
-                      "slideshow",  NULL };
+                      "slideshow",  "print", NULL };
```
#ifdef HAVE_PRINT
         "print", 
#endif

TurboGit replied:
Indeed, fixed too.

TurboGit commented:
Thanks for all the comments. I have integrated all of them and pushed a new version.

houz commented:
Could you merge in master so that I can see what I actually have to port to GTK3?

TurboGit commented:
Now merged. Thanks for looking at this. I'll continue working on the print-module after.

boucman commented:
ok, @hanatos meant to merge master into print-module, not the other way round

I have reverted the merge and rebased the print-module branch, hopefully I got it right.

feel free to reopen the PR and continue discussion there

TurboGit commented:
Le dimanche 04 janvier 2015 à 08:23 -0800, Boucman a écrit :
> ok, @hanatos meant to merge master into print-module, not the other
> way round
> 
> I have reverted the merge and rebased the print-module branch,
> hopefully I got it right.
> 
> feel free to reopen the PR and continue discussion there

Ah! ok, sorry :( Thanks for reverting.

-- 
  Pascal Obry /  Magny Les Hameaux (78)

  The best way to travel is by means of imagination

  http://v2p.fr.eu.org
  http://www.obry.net

  gpg --keyserver keys.gnupg.net --recv-key F949BD3B

TurboGit commented:
Note that to my defense the message said "merge in master" and I thought it was fine because one can always disable this module.

Anyway, I don't see a way to reopen this PR?
